### ISSUE TITLE
Critically analyse every PV we set on the eiger

### ISSUE THREAD

**Issue by DominicOram** [DiamondLightSource/python-artemis#287]
Artemis currently sets forward stream to true with no justification other than doing what GDA does. We should be more careful around what we're setting and why, not just do it because GDA does, particularly in the eiger/odin where there's a lot of PVs set

# Acceptance Criteria
* Artemis doesn't set the forward stream
* We critically analyse the PVs set and document with comments why they're required

**Comment by DominicOram** [comment]
FYI @graeme-winter

**Comment by DominicOram** [comment]
Some thoughts in reading the code that may need further discussion: 
* When we change ROI we set `NumRowChunks` `NumColChunks` to match detector height/width. This seems sensible to me but we also set the `NumFramesChunks` to 1 at the same time. This may be a sensible value but seems unrelated to ROI from my understanding of chunking
* We set the trigger mode to External Series and then the number of triggers to 1. From (https://areadetector.github.io/master/ADEiger/eiger.html) I believe the same behavior can be achieved by just setting the mode to External Enable. I don't think this makes much difference other than one less PV setting might be slightly less time to wait on stale params
* Before we arm the detector we're reading the bit depth out of the eiger and pushing it to the filewriter. This seems a bit of an odd responsibility for Artemis to have, can it not be done at the EPICS layer?
* For the detector photon energy we check if it's close to the requested energy before setting it again, we don't do this for any other PVs. If we're doing this to save time why are we only doing it here?

**Comment by graeme-winter** [comment]
> * When we change ROI we set `NumRowChunks` `NumColChunks` to match detector height/width. This seems sensible to me but we also set the `NumFramesChunks` to 1 at the same time. This may be a sensible value but seems unrelated to ROI from my understanding of chunking

This should be the only chunk size we ever use -> 1 image (this is mandated by the way the data are sent pre-compressed from the DCU)

> * We set the trigger mode to External Series and then the number of triggers to 1. From (https://areadetector.github.io/master/ADEiger/eiger.html) I believe the same behavior can be achieved by just setting the mode to External Enable. I don't think this makes much difference other than one less PV setting might be slightly less time to wait on stale params

Am less offended by being explicit here

> * Before we arm the detector we're reading the bit depth out of the eiger and pushing it to the filewriter. This seems a bit of an odd responsibility for Artemis to have, can it not be done at the EPICS layer?

This is a "feature" of how ODIN works - strictly it _should_ pick up the bit depth from the header packet - `wyrmberg` did this - but ODIN needs to be pre-warned. We do not have a way of determining the depth _without_ asking the Eiger. I recognise that this is untidy.

> * For the detector photon energy we check if it's close to the requested energy before setting it again, we don't do this for any other PVs. If we're doing this to save time why are we only doing it here?

Because changing the photon energy changes the threshold which I think is slow? However how frequently is that an issue is something I cannot answer.

**Comment by npat113** [comment]
threshold changes should be rare - XRC should use the current energy and scale accordingly (handled in xraycentring.py).

**Comment by npat113** [comment]
It would be useful to get an idea of how much time is spent in odin PV setting. I do wonder if sending info directly to the detector and not using Odin would be a more efficient way of operating.

**Comment by DominicOram** [comment]
> This should be the only chunk size we ever use -> 1 image (this is mandated by the way the data are sent pre-compressed from the DCU)

Great, I'll move it out of ROI and put it somewhere more generic

>> * Before we arm the detector we're reading the bit depth out of the eiger and pushing it to the filewriter. This seems a bit of an odd responsibility for Artemis to have, can it not be done at the EPICS layer?

> This is a "feature" of how ODIN works - strictly it should pick up the bit depth from the header packet - wyrmberg did this - but ODIN needs to be pre-warned. We do not have a way of determining the depth without asking the Eiger. I recognise that this is untidy.

This is what I thought, fair enough. @GDYendell is there an issue or something documenting this requirement for odin that I can point a comment to?

> Because changing the photon energy changes the threshold which I think is slow? However how frequently is that an issue is something I cannot answer.
>threshold changes should be rare - XRC should use the current energy and scale accordingly (handled in xraycentring.py).
so basically we're not switching threshold back and forth.

> It would be useful to get an idea of how much time is spent in odin PV setting. I do wonder if sending info directly to the detector and not using Odin would be a more efficient way of operating.

Ok, more timing analysis etc. required. Documented at https://github.com/DiamondLightSource/python-artemis/issues/293

**Comment by GDYendell** [comment]
> is there an issue or something documenting this requirement

Not as well as I would like. The best documentation of this is that it is what test acquisition runner does: https://github.com/dls-controls/ADOdin/blob/dls-master/etc/tools/odin_acquisition/odin_acquisition/eiger_acquisition.py#L95

**Comment by npat113** [comment]
Dom, bit depth is set based on frame rate. I don't think there is an energy  or ROI dependence, but we can easily check. I suspect it will be the same for every XRC collection unless someone decides to use 10ms+ exposure.

**Comment by bentom08** [comment]
re: the threshold changes - this is the ticket in which we starting checking the difference from the current threshold in GDA
https://jira.diamond.ac.uk/browse/I03-363

**Comment by npat113** [comment]
Am I correct in thinking that the detector can accept all parameters at once via a JSON payload? Would it be more efficient to bypass epics and odin for pushing the parameters across and simply have them read them once it is all done?

**Comment by GDYendell** [comment]
Not that I am aware of. I had a guess at what that API might be and it doesn't seem to like it:
```
❯ curl -X PUT -d '{"value":"test"}' -H 'Content-Type: application/json' http://i07-eiger01/stream/api/1.8.0/config/header_appendix
❯ curl -H 'Content-Type: application/json' http://i07-eiger01/stream/api/1.8.0/config/header_appendix
{"access_mode":"rw","value":"test","value_type":"string"}
❯ curl -X PUT -d '{"value":{"header_detail": "all", "header_appendix": "test2"}}' -H 'Content-Type: application/json' http://i07-eiger01/stream/api/1.8.0/config
error during request: path error
```
Is that something you have done before @npat113 ?

As for whether it would make a significant difference I am not sure.

**Comment by graeme-winter** [comment]
https://media.dectris.com/210607-DECTRIS-SIMPLON-API-Manual_EIGER2-chip-based_detectros.pdf

**Comment by graeme-winter** [comment]
"Meta information in the body of the request and in the reply from the HTTP server are serialized in the JSON format and described in the table below."

P16 of PDF - following pages include e.g. Python `requests` example of how to pass information. I am fairly sure that places-that-are-not-here routinely use the full JSON configuration blob to set the detector up all in one go.

**Comment by GDYendell** [comment]
That example is only setting one parameter. Am I looking at the wrong thing?

**Comment by graeme-winter** [comment]
> That example is only setting one parameter. Am I looking at the wrong thing?

OK, if I can find a chance I will have a go at poking an Eiger with > 1 - though guess one on i03 is off right now?

**Comment by GDYendell** [comment]
In any case, it may be a bad idea to bypass the IOC because the only way the IOC updates a lot of readbacks is by fetching the parameters returned from PUTs. If it is not doing the PUT it won't fetch the changes.

Could we profile it (once Artemis is doing put with callback to avoid the race condition) and see what the overhead is?

**Comment by DominicOram** [comment]
Yes, we can profile as part of https://github.com/DiamondLightSource/python-artemis/issues/293. Thanks for looking all, I think my questions on this ticket have been addressed, with further details requiring profiling to be done in 293.
